src/devices/src/virtio/block/device.rs

```diff
@@ -215,6 +216,13 @@ impl Block {
 
         let image_type = disk::detect_image_type(&disk_image, false).unwrap();
 
+        if image_type != disk_image_format {
+            return Err(io::Error::new(
+                ErrorKind::InvalidInput,
+                "image type reported does not match image header",
+            ));
+        }
+
```

Comment:
I would have expected the user’s choice overrides all possible probing. I.e., if the user specifies an image as raw, it should always be opened as raw regardless of whether there’s a qcow2 magic in the first four bytes.

Reply:
Should be fixed.